google/cloud/bigquery/_helpers.py

```diff
@@ -1004,3 +1005,61 @@ def _verify_job_config_type(job_config, expected_type, param_name="job_config"):
                 job_config=job_config,
             )
         )
+
+
+class ResourceBase:
```

Comment:
I saw the discussion between you and Tim, but on a second thought, I feel like it might be more prudent to keep the boilerplate code instead of using mixin. It indeed is some duplicate code, but I think
- they are very readable duplicate code
- we avoid any unintended consequence of multiple inheritance, for example, if the child class by accident has a method or property with the same name as the parent class, it may cause mysterious behavior and become very hard to debug
- we ensure each resource class is strictly self-contained
- it's easier to maintain, for example, future programmers don't have to dig into the parent class to find where `self._properties` is
- having `def from_api_repr():` present in each resource class reminds people to write unit tests for it, and coverage test will report an error if we didn't

Overall I feel like the benefit is much outweighed by the cost. This is just a preference of mine, and I would be glad to hear your opinion on this and discuss more!

Reply:
I hear you on the desire to avoid inheritance where we can. Good point.

I have several desires/hopes for our work in BQ over time.
* Improve the functions as a whole with docstrings, etc.
* Reduce duplicate code where appropriate

If we pursue the idea of having a docstring for a function like this (which arguably may OR may not be necessary), we are looking at something akin to 10+ lines of code in a function that shows up ~60 times for about 600+ lines of code if we universally update all the docstrings now that we have one that is viable.

Another alternative is to create a helper function and then reference the helper function in our classes (NOTE: THIS IS PSEUDOCODE ... I have not actually tested all this. Just posting something to facilitate discussion. The actual code will likely be slightly different):

```
def _from_api_repr(obj, resource: dict):
    # Add docstring here ...
    config = obj
    config._properties = copy.deepcopy(resource)
    return config

from google.cloud.bigquery._helpers import _from_api_repr

class <ClassName>:
    def from_api_repr(self, resource):
        return = _from_api_repr(self, resource)  # Pass self to function as an arg.
```

The benefit of a helper function:
* No class inheritance
* Trivial to override for specific use cases
* Reduces some duplication in the Docstring area (I am aware it doesn't really save much in terms of reducing the amount of code duplication in each of the classes)

### Function in question:
```
@classmethod
def from_api_repr(cls, resource: dict):
    """Factory: constructs an instance of the class (cls)
    given its API representation.
    Args:
        resource (Dict[str, Any]):
            API representation of the object to be instantiated.
    Returns:
        ResourceBase: An instance of the class initialized with data
            from 'resource'.
    """
    config = cls()
    config._properties = copy.deepcopy(resource)
    return config     
```

Reply:
I think the idea of a helper function is a good balance. It's clear that the code is duplicated, and adds the clarity of having the function be present in the class.